File: mobile-sdk/android/FmcwsSDK.kt
/**
 * FMCWS Android SDK v1.0.0
 * Файл: FmcwsSDK.kt
 *
 * Размещение: fmcws-android-sdk/src/main/java/by/fmcws/sdk/FmcwsSDK.kt
 *
 * Подключение в build.gradle (app):
 *   implementation(files("libs/fmcws-sdk-1.0.0.aar"))
 *   implementation("com.squareup.okhttp3:okhttp:4.12.0")
 *   implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.8.0")
 *
 * Инициализация (Application.onCreate или MainActivity):
 *   FmcwsSDK.init(FmcwsConfig(
 *       apiUrl  = "https://fmcws.bank.by",
 *       apiKey  = "YOUR_API_KEY",
 *       clientId = "PSEUDONYMIZED_CLIENT_ID"
 *   ))
 *
 * Отправка события:
 *   FmcwsSDK.track("TRANSFER_CREATE", mapOf("amount" to 1500.0, "currency" to "BYN"))
 */

package by.fmcws.sdk

import android.content.Context
import android.os.Build
import android.provider.Settings
import android.util.Log
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.security.MessageDigest
import java.util.UUID
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit

// ── Config ────────────────────────────────────────────────────────────────────
data class FmcwsConfig(
    val apiUrl: String,
    val apiKey: String,
    val clientId: String,
    val debug: Boolean = false,
    val flushIntervalMs: Long = 5_000L,     // буферизация событий
    val maxQueueSize: Int = 100,
)

// ── Warning levels ─────────────────────────────────────────────────────────────
enum class FmcwsRiskLevel { LOW, MEDIUM, HIGH, CRITICAL }

data class FmcwsWarning(
    val level: FmcwsRiskLevel,
    val message: String,
    val recommendation: String,
)

// ── Warning callback ───────────────────────────────────────────────────────────
interface FmcwsWarningListener {
    fun onWarning(warning: FmcwsWarning)
}

// ── SDK Singleton ──────────────────────────────────────────────────────────────
object FmcwsSDK {
    private const val TAG = "FMCWS"
    private const val PREF_SESSION = "fmcws_session_id"

    private lateinit var config: FmcwsConfig
    private lateinit var context: Context
    private lateinit var sessionId: String
    private lateinit var deviceId: String

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val queue = LinkedBlockingQueue<JSONObject>(500)
    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .writeTimeout(10, TimeUnit.SECONDS)
        .build()

    private var warningListener: FmcwsWarningListener? = null
    private var initialized = false

    // ── Инициализация ──────────────────────────────────────────────────────────
    fun init(appContext: Context, cfg: FmcwsConfig) {
        if (initialized) return
        context = appContext.applicationContext
        config = cfg

        sessionId = getOrCreateSessionId()
        deviceId  = getDeviceId()

        // Запуск фонового воркера отправки
        startFlushWorker()
        initialized = true
        log("SDK initialized | session=$sessionId | device=$deviceId")
    }

    fun setWarningListener(listener: FmcwsWarningListener) {
        warningListener = listener
    }

    // ── Публичное API ──────────────────────────────────────────────────────────
    fun track(eventType: String, extra: Map<String, Any?> = emptyMap()) {
        checkInit()
        val payload = buildPayload(eventType, extra)
        if (queue.size >= config.maxQueueSize) {
            log("Queue full, dropping oldest event")
            queue.poll()
        }
        queue.offer(payload)
        log("Queued: $eventType (queue size=${queue.size})")
    }

    /** Обновить clientId после входа в аккаунт */
    fun identify(clientId: String) {
        config = config.copy(clientId = clientId)
    }

    /** Сбросить сессию (выход из аккаунта) */
    fun reset() {
        context.getSharedPreferences("fmcws", Context.MODE_PRIVATE)
            .edit().remove(PREF_SESSION).apply()
        sessionId = createNewSession()
        log("Session reset: $sessionId")
    }

    // ── Payload builder ────────────────────────────────────────────────────────
    private fun buildPayload(eventType: String, extra: Map<String, Any?>): JSONObject {
        val extraJson = JSONObject()
        extra.forEach { (k, v) -> extraJson.put(k, v) }

        return JSONObject().apply {
            put("session_id",          sessionId)
            put("client_id",           config.clientId)
            put("event_type",          eventType)
            put("channel",             "android")
            put("device_fingerprint",  deviceId)
            put("user_agent",          buildUserAgent())
            put("extra",               extraJson)
        }
    }

    private fun buildUserAgent(): String {
        return "FmcwsAndroidSDK/1.0.0 " +
               "Android/${Build.VERSION.RELEASE} " +
               "SDK/${Build.VERSION.SDK_INT} " +
               "${Build.MANUFACTURER}/${Build.MODEL}"
    }

    // ── Flush worker ───────────────────────────────────────────────────────────
    private fun startFlushWorker() {
        scope.launch {
            while (isActive) {
                delay(config.flushIntervalMs)
                flush()
            }
        }
    }

    private suspend fun flush() {
        while (queue.isNotEmpty()) {
            val payload = queue.peek() ?: break
            val success = sendEvent(payload)
            if (success) {
                queue.poll()
            } else {
                // Не удаляем из очереди — повторим на следующем цикле
                delay(3_000)
                break
            }
        }
    }

    // ── HTTP отправка ──────────────────────────────────────────────────────────
    private suspend fun sendEvent(payload: JSONObject): Boolean = withContext(Dispatchers.IO) {
        val url = config.apiUrl.trimEnd('/') + "/v1/events"
        val body = payload.toString().toRequestBody("application/json".toMediaType())
        val request = Request.Builder()
            .url(url)
            .post(body)
            .header("X-API-Key", config.apiKey)
            .header("X-SDK-Version", "android-1.0.0")
            .build()

        return@withContext try {
            client.newCall(request).execute().use { response ->
                if (response.isSuccessful) {
                    val json = JSONObject(response.body?.string() ?: "{}")
                    handleResponse(json)
                    true
                } else {
                    log("HTTP error: ${response.code}")
                    false
                }
            }
        } catch (e: IOException) {
            log("Network error: ${e.message}")
            false
        }
    }

    // ── Обработка ответа (предупреждение) ──────────────────────────────────────
    private fun handleResponse(json: JSONObject) {
        val showWarning = json.optBoolean("show_warning", false)
        if (!showWarning) return

        val level = when (json.optString("risk_level")) {
            "critical" -> FmcwsRiskLevel.CRITICAL
            "high"     -> FmcwsRiskLevel.HIGH
            "medium"   -> FmcwsRiskLevel.MEDIUM
            else       -> FmcwsRiskLevel.LOW
        }
        val message = json.optString("warning_message", "")
        if (message.isBlank()) return

        val warning = FmcwsWarning(level, message, json.optString("recommendation", ""))
        log("Warning received: $level — $message")

        // Callback в main thread
        CoroutineScope(Dispatchers.Main).launch {
            warningListener?.onWarning(warning)
        }
    }

    // ── Device ID (псевдонимизированный) ──────────────────────────────────────
    private fun getDeviceId(): String {
        val androidId = Settings.Secure.getString(
            context.contentResolver, Settings.Secure.ANDROID_ID
        ) ?: UUID.randomUUID().toString()

        // SHA-256 для псевдонимизации
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(("fmcws_$androidId").toByteArray())
        return "adv_" + hash.take(8).joinToString("") { "%02x".format(it) }
    }

    // ── Session ID ─────────────────────────────────────────────────────────────
    private fun getOrCreateSessionId(): String {
        val prefs = context.getSharedPreferences("fmcws", Context.MODE_PRIVATE)
        return prefs.getString(PREF_SESSION, null) ?: createNewSession()
    }

    private fun createNewSession(): String {
        val sid = UUID.randomUUID().toString()
        context.getSharedPreferences("fmcws", Context.MODE_PRIVATE)
            .edit().putString(PREF_SESSION, sid).apply()
        return sid
    }

    private fun checkInit() {
        check(initialized) { "[FMCWS] Call FmcwsSDK.init() before tracking events" }
    }

    private fun log(msg: String) {
        if (config.debug) Log.d(TAG, msg)
    }
}

File: mobile-sdk/ios/FmcwsSDK.swift
/**
 * FMCWS iOS SDK v1.0.0
 * Файл: FmcwsSDK.swift
 *
 * Размещение: FMCWS.xcframework (встраивается в проект банка)
 *
 * Подключение (Package.swift / CocoaPods / прямое встраивание XCFramework):
 *   Скопировать FMCWS.xcframework в проект → General → Frameworks → Embed & Sign
 *
 * Swift Package Manager (Package.swift):
 *   .binaryTarget(name: "FMCWS", path: "FMCWS.xcframework")
 *
 * Инициализация (AppDelegate.swift или @main):
 *   FmcwsSDK.shared.initialize(
 *     config: FmcwsConfig(
 *       apiURL:   URL(string: "https://fmcws.bank.by")!,
 *       apiKey:   "YOUR_API_KEY",
 *       clientId: "PSEUDONYMIZED_CLIENT_ID"
 *     )
 *   )
 *
 * Трекинг события:
 *   FmcwsSDK.shared.track("TRANSFER_CREATE", extra: ["amount": 1500.0])
 */

import Foundation
import UIKit
import CryptoKit

// ── Config ────────────────────────────────────────────────────────────────────
public struct FmcwsConfig {
    public let apiURL:         URL
    public let apiKey:         String
    public var clientId:       String
    public let debug:          Bool
    public let flushInterval:  TimeInterval   // секунды между flush (default 5)
    public let maxQueueSize:   Int

    public init(
        apiURL:        URL,
        apiKey:        String,
        clientId:      String,
        debug:         Bool         = false,
        flushInterval: TimeInterval = 5.0,
        maxQueueSize:  Int          = 100
    ) {
        self.apiURL        = apiURL
        self.apiKey        = apiKey
        self.clientId      = clientId
        self.debug         = debug
        self.flushInterval = flushInterval
        self.maxQueueSize  = maxQueueSize
    }
}

// ── Risk / Warning ─────────────────────────────────────────────────────────────
public enum FmcwsRiskLevel: String {
    case low, medium, high, critical
}

public struct FmcwsWarning {
    public let level:          FmcwsRiskLevel
    public let message:        String
    public let recommendation: String
}

public protocol FmcwsWarningDelegate: AnyObject {
    func fmcws(_ sdk: FmcwsSDK, didReceiveWarning warning: FmcwsWarning)
}

// ── SDK ────────────────────────────────────────────────────────────────────────
public final class FmcwsSDK {

    // MARK: – Singleton
    public static let shared = FmcwsSDK()
    private init() {}

    // MARK: – State
    private var config: FmcwsConfig?
    private var sessionId: String = ""
    private var deviceId:  String = ""

    private var queue: [[String: Any]] = []
    private let queueLock = NSLock()
    private var flushTimer: Timer?

    private let session = URLSession(configuration: {
        let c = URLSessionConfiguration.default
        c.timeoutIntervalForRequest  = 10
        c.timeoutIntervalForResource = 10
        return c
    }())

    public weak var warningDelegate: FmcwsWarningDelegate?

    // MARK: – Init
    public func initialize(config: FmcwsConfig) {
        guard self.config == nil else { return }
        self.config    = config
        self.sessionId = getOrCreateSessionId()
        self.deviceId  = makeDeviceId()
        startFlushTimer()
        log("SDK initialized | session=\(sessionId) | device=\(deviceId)")
    }

    // MARK: – Public API

    /// Отправить событие (неблокирующий — кладёт в очередь)
    public func track(_ eventType: String, extra: [String: Any] = [:]) {
        guard let config = config else {
            assertionFailure("[FMCWS] Call initialize(config:) first")
            return
        }
        let payload: [String: Any] = [
            "session_id":         sessionId,
            "client_id":          config.clientId,
            "event_type":         eventType,
            "channel":            "ios",
            "device_fingerprint": deviceId,
            "user_agent":         makeUserAgent(),
            "extra":              extra,
        ]
        queueLock.lock()
        if queue.count >= (config.maxQueueSize) { queue.removeFirst() }
        queue.append(payload)
        queueLock.unlock()
        log("Queued: \(eventType) (queue=\(queue.count))")
    }

    /// Обновить clientId после авторизации
    public func identify(_ clientId: String) {
        config?.clientId = clientId
    }

    /// Сбросить сессию при выходе
    public func reset() {
        UserDefaults.standard.removeObject(forKey: "fmcws_session_id")
        sessionId = createNewSession()
        log("Session reset: \(sessionId)")
    }

    // MARK: – Flush

    private func startFlushTimer() {
        let interval = config?.flushInterval ?? 5.0
        DispatchQueue.main.async {
            self.flushTimer = Timer.scheduledTimer(
                withTimeInterval: interval, repeats: true
            ) { [weak self] _ in self?.flush() }
        }
    }

    private func flush() {
        queueLock.lock()
        guard !queue.isEmpty else { queueLock.unlock(); return }
        let payload = queue.first!
        queueLock.unlock()

        sendEvent(payload) { [weak self] success in
            guard let self = self else { return }
            if success {
                self.queueLock.lock()
                if !self.queue.isEmpty { self.queue.removeFirst() }
                self.queueLock.unlock()
                // Рекурсивно отправляем оставшиеся
                if !self.queue.isEmpty { self.flush() }
            }
        }
    }

    // MARK: – HTTP

    private func sendEvent(_ payload: [String: Any], completion: @escaping (Bool) -> Void) {
        guard let config = config else { return }
        let url = config.apiURL.appendingPathComponent("/v1/events")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(config.apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("ios-1.0.0", forHTTPHeaderField: "X-SDK-Version")

        guard let body = try? JSONSerialization.data(withJSONObject: payload) else {
            completion(false); return
        }
        request.httpBody = body

        session.dataTask(with: request) { [weak self] data, response, error in
            if let error = error {
                self?.log("Network error: \(error.localizedDescription)")
                completion(false); return
            }
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                self?.log("HTTP error")
                completion(false); return
            }
            if let data = data,
               let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                self?.handleResponse(json)
            }
            completion(true)
        }.resume()
    }

    // MARK: – Response handling

    private func handleResponse(_ json: [String: Any]) {
        guard json["show_warning"] as? Bool == true else { return }
        let levelStr = json["risk_level"] as? String ?? "low"
        let level: FmcwsRiskLevel = FmcwsRiskLevel(rawValue: levelStr) ?? .low
        guard level != .low, let message = json["warning_message"] as? String, !message.isEmpty else { return }

        let warning = FmcwsWarning(
            level:          level,
            message:        message,
            recommendation: json["recommendation"] as? String ?? ""
        )
        log("Warning: \(level.rawValue) — \(message)")
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.warningDelegate?.fmcws(self, didReceiveWarning: warning)
        }
    }

    // MARK: – Device ID (псевдонимизированный SHA-256)

    private func makeDeviceId() -> String {
        let raw = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
        let salted = "fmcws_\(raw)".data(using: .utf8)!
        let hash = SHA256.hash(data: salted)
        let hex = hash.prefix(8).map { String(format: "%02x", $0) }.joined()
        return "idfv_\(hex)"
    }

    // MARK: – Session

    private func getOrCreateSessionId() -> String {
        if let sid = UserDefaults.standard.string(forKey: "fmcws_session_id") { return sid }
        return createNewSession()
    }

    private func createNewSession() -> String {
        let sid = UUID().uuidString
        UserDefaults.standard.set(sid, forKey: "fmcws_session_id")
        return sid
    }

    // MARK: – User-Agent

    private func makeUserAgent() -> String {
        let info = Bundle.main
        let appName    = info.infoDictionary?["CFBundleName"] as? String ?? "BankApp"
        let appVersion = info.infoDictionary?["CFBundleShortVersionString"] as? String ?? "?"
        let osVersion  = UIDevice.current.systemVersion
        let model      = UIDevice.current.model
        return "FmcwsiOSSDK/1.0.0 \(appName)/\(appVersion) iOS/\(osVersion) \(model)"
    }

    private func log(_ msg: String) {
        if config?.debug == true { print("[FMCWS] \(msg)") }
    }
}
